File: Harness Typescript/src/evaluators/json-schema.ts
import Ajv from 'ajv';
import { BaseEvaluator, EvaluatorResult } from './base.js';

export class JsonSchemaEvaluator extends BaseEvaluator {
  private ajv: any;

  constructor() {
    super("json_schema");
    const AjvConstructor = (Ajv as any).default || Ajv;
    this.ajv = new AjvConstructor();
  }

  async evaluate(output: string, expected: any): Promise<EvaluatorResult> {
    try {
      // Find JSON block if it's wrapped in markdown ```json
      let cleanOutput = output;
      const match = output.match(/```(?:json)?\s*([\s\S]*?)```/);
      if (match) {
        cleanOutput = match[1];
      }

      const parsed = JSON.parse(cleanOutput);
      const validate = this.ajv.compile(expected);
      const valid = validate(parsed);

      if (!valid) {
        return {
          passed: false,
          score: 0.0,
          reason: `Schema validation error: ${this.ajv.errorsText(validate.errors)}`
        };
      }
      return { passed: true, score: 1.0 };
    } catch (e) {
      return {
        passed: false,
        score: 0.0,
        reason: `Invalid JSON format: ${(e as Error).message}`
      };
    }
  }
}

File: Harness Python/evaluators/json_schema.py
import json
import jsonschema
from evaluators.base import BaseEvaluator, EvalResult

class JsonSchemaEvaluator(BaseEvaluator):
    async def evaluate(self, output: str, test_case: dict) -> EvalResult:
        schema = test_case.get("schema", {})
        if not schema:
            return EvalResult(passed=False, score=0.0, reason="No schema provided")
            
        # Strip markdown fences if any
        clean_output = output.strip()
        if clean_output.startswith("```json"):
            clean_output = clean_output[7:]
        elif clean_output.startswith("```"):
            clean_output = clean_output[3:]
        if clean_output.endswith("```"):
            clean_output = clean_output[:-3]
            
        clean_output = clean_output.strip()
        
        try:
            parsed = json.loads(clean_output)
            jsonschema.validate(instance=parsed, schema=schema)
            return EvalResult(passed=True, score=1.0)
        except json.JSONDecodeError as e:
            return EvalResult(passed=False, score=0.0, reason=f"Invalid JSON: {e}")
        except jsonschema.ValidationError as e:
            return EvalResult(passed=False, score=0.0, reason=f"Schema validation failed: {e.message}")
